No indication of failed data gathering from Hawkular endpoint

yaacov opened the pull request:
**Description**
When gathering hawkular metrics fails, user has no indication in the status ui page.

This PR updates the status of the Metrics Collection Status line each time metrics is collected.
It also display the last update date and time as an html title for each endpoint status, so users can know when the status was last checked.

**Screenshots**
![metrics_valid](https://cloud.githubusercontent.com/assets/2181522/17293840/7901f13a-57fb-11e6-9c25-f0920bb61328.png)

![metrics_invalid](https://cloud.githubusercontent.com/assets/2181522/17293842/7d8534c4-57fb-11e6-8636-79b988907816.png)

**Bugzzila**
https://bugzilla.redhat.com/show_bug.cgi?id=1344308
https://bugzilla.redhat.com/show_bug.cgi?id=1356130

yaacov commented:
@miq-bot add_label providers/containers,bug

@simon3z @moolitayer @zeari please review

simon3z commented:
@blomquisg is this really demanded to each provider? Do you have/plan something in the infrastructure? (I guess that the challenge there is to identify what was the relevant endpoint that failed).

chessbyte commented:
@simon3z - @blomquisg is on vacation for several weeks.  Is this urgent?

simon3z commented:
> @simon3z - @blomquisg is on vacation for several weeks. Is this urgent?

@chessbyte it's not urgent (if it's 1-2 weeks), I have the feeling that the infra should be involved in this, who else can reply here or design the missing part (if any). Jason?

yaacov commented:
@miq-bot add_label darga/yes

https://bugzilla.redhat.com/show_bug.cgi?id=1356130

chessbyte commented:
@dclarizio would you please provide your input on an approach for this issue?

simon3z commented:
@blomquisg is you want to merge this and provide the infrastructure later is fine by me. Let's just not forget that we have to clean this up.

blomquisg commented:
@simon3z agreed .. this is a good fix, but definitely something we should do across the board.

blomquisg commented:
@yaacov @simon3z Looking at this again, I think there needs to be a piece that ties into the Schedule Worker to keep the status of the authentication row up to date even if the Metrics worker dies.

Look at [RHEV's `authentications_to_validate` method](https://github.com/ManageIQ/manageiq/blob/d42d31d8014f3196b003d989ba42ce7ba2e9c940/app/models/manageiq/providers/redhat/infra_manager.rb#L161).  If you implement that method in the container managers, then the Schedule Worker will validate the status of the authentications on a regular basis and not rely on the Metrics worker to be running.

yaacov commented:
>  to keep the status of the authentication row up to date even if the Metrics worker dies

The bug is about "**No indication of failed data gathering**"
In containers we check the authentication validation by trying to read metrics ... so it was easy to indicate the metrics reading status by showing the user the authentication state.
Thinking about it again it does sound confusing :-(
@simon3z @blomquisg WDYT ?

simon3z commented:
> Thinking about it again it does sound confusing :-(

@yaacov what part is confusing?

yaacov commented:
> what part is confusing?

@simon3z we check for metrics collection but the see the title in the status card is "Hawkular Authentication" 

It's not a big problem :-)
1. I will add a new line in the status card with a title ~"Metric Collection" with data about metric collection
2. then I can use @blomquisg suggestion and update the bearer and default authentications without spoiling the metric collection information.

simon3z commented:
@blomquisg in the other providers you have "Authentication Status" but it actually catches any kind of exception? (E.g. connection failed, etc.?) or is it really scoped to just authentication failures?

yaacov commented:
@simon3z Changes made:

1. The bearer and hawkular authentications are not touched by the metrics collection.
2. added authentications_to_validate method like @blomquisg suggested.
3. added a new line in the status with title "metrics collection"

yaacov commented on app/models/manageiq/providers/kubernetes/container_manager/metrics_capture.rb:
```diff
@@ -37,8 +37,19 @@ class TargetValidationError < RuntimeError; end
+      return unless ems
+
+      auth = ems.authentication_type(:metrics_collection)
+      ems.authentications << Authentication.new(:authtype => :metrics_collection) unless auth
```
Adding an authentication here is not a nice thing to do.

**We can also:**
a. Create the authentication wile creating a new provider.
b. Add new fields to the provider class.

**Pros:**
1. We support providers created without this authentication - no need to edit the provider and add this authentication to all the previously created providers. 
2. We do not need new fields in the ems class - no need for migrations.
**Cons:**
1. Creating an authentication while reading data - bad thing to do.


@moolitayer @alongoldboim @simon3z help please ?

moolitayer replied:
I'm trying to understand if this (creating an authentication during capture) was what @simon3z and @blomquisg agreed to include and fix later. If so lets add a comment and proceed with it.

Anyway I would think on master we will need to replace it with creating the endpoint on provider addition + migration to handle old providers. Lets discuss this tomorrow.

yaacov replied:
@moolitayer 
> [ is this ] ... what @simon3z and @blomquisg agreed ... ?

No, It is the best solution I found so far, It does need to be discussed, and may be a better solution will show itself ...

moolitayer commented on app/helpers/ems_container_helper/textual_summary.rb:
```diff
@@ -19,25 +20,17 @@ def textual_group_relationships
 
   def textual_group_status
-    textual_authentications + %i(refresh_status)
+    textual_authentications + %i(metrics_status) + %i(refresh_status)
```
```textual_authentications +  %i(metrics_status, refresh_status)``` ?

yaacov replied:
:+1:  Thanks ! missed that.

yaacov commented:
@simon3z @moolitayer now using migrations, please review.

enoodle commented on app/helpers/textual_mixins/textual_metrics_status.rb:
```diff
@@ -0,0 +1,39 @@
+    end
+  end
+
+  def textual_authentication_value(status, updated_on, valid_str = _("Valid"))
```
Are we sure that this is the right place for this function? It seems good and helpful but I am not sure that we want to keep it in this file / module. Same goes for ```textual_authentication_value```. Maybe we should have a ```textual_authentication_status``` file as well?

yaacov replied:
:+1:

enoodle commented:
LGTM :+1:

simon3z commented:
@blomquisg this is becoming infrastructural, do you want to take a look?

blomquisg commented on app/helpers/textual_mixins/textual_authentications_status.rb:
```diff
@@ -0,0 +1,41 @@
+
+  def textual_authentications_status
+    authentications = [@ems.authentication_type(:bearer), @ems.authentication_type(:hawkular)].compact
+    return [{:label => _("Bearer Authentication"), :title => t = _("None"), :value => t}] if authentications.blank?
```
I really like this PR, except for this one bit.

Is there a way to keep this out of this file?  Or, even abstracted away inside the file to allow different providers to use the same `textual_` method?

Maybe the `@ems` can report the list of `authentication_types` and that can drive the list of `authentications` here.

yaacov replied:
:+1:  It's now more generic

blomquisg commented:
> @blomquisg in the other providers you have "Authentication Status" but it actually catches any kind of exception? (E.g. connection failed, etc.?) or is it really scoped to just authentication failures?

Unfortunately, it's probably a mixed bag actually.  Another area that pluggability could help.

blomquisg commented:
> @blomquisg this is becoming infrastructural, do you want to take a look?

That's definitely true.  I would like all the other providers to present this same data in the summary page.  Basically, all endpoints of a provider should show an authentication status.  Or, really a connection/authentication status.  But, baby steps are fine. :)

So, if https://github.com/ManageIQ/manageiq/pull/9785#discussion_r74146530 can be addressed, I think that would essentially make this a purely infrastructural change with a first implementation for Containers.

yaacov commented:
@simon3z @blomquisg Updated the authentications list to be more generic.

simon3z commented on db/migrate/20160808150745_add_metrics_status_to_ext_management_system.rb:
```diff
@@ -0,0 +1,7 @@
+  def change
+    add_column :ext_management_systems, :last_metrics_error, :text
+    add_column :ext_management_systems, :last_metrics_update_date, :timestamp
+    add_column :ext_management_systems, :last_metrics_success_date, :timestamp
```
This doesn't seems generic enough. @blomquisg I guess you would like to have a status per endpoint. Not something hard-coded for metrics.

blomquisg replied:
@simon3z yeah ... man, I need to think about this one.

Lemme see if I can break this down a bit.  We currently have:

* `authentications.status`: stores the status of the authentication/connection
  - `authentications.status_details` has the details of the error, if there was one
* `ext_management_systems.last_refresh_date`: stores the date of the last ems refresh
  - `ext_management_systems.last_refresh_error` has the details of the last refresh failure, if there was one.

We don't currently store any type of status on the `endpoints` table.

Also, part of the problem is that we don't store anything in the DB specifying what ManageIQ Provider Capability each each endpoint is used for (i.e., Refresh, Events, Metrics).  That distinction could be captured just in the logic of each Manager.  But, if a single endpoint is used for multiple capabilities (e.g., a single endpoint used for Refresh *and* Metrics) there would need to be some way to indicate the different status values for each capability on that endpoint.

I don't see us figuring that out anytime soon, honestly.

For now it might be better to just add the metrics status columns to `ext_management_systems` and refactor the status tracking later.

moolitayer replied:
:+1: for going with this solution until we have something better.
I suggest the name `last metrics_success`

blomquisg replied:
I think `last_metrics_status`, since it holds ["Valid"](https://github.com/ManageIQ/manageiq/pull/9785/files#diff-189fb4a1163fabfb715cd145f5da6edaR68) as well as failure states like ["Unavailable"](https://github.com/ManageIQ/manageiq/pull/9785/files#diff-189fb4a1163fabfb715cd145f5da6edaR62).

yaacov replied:
@simon3z @moolitayer @blomquisg Update:
`last_metrics_error` column now only holds the error, or `nil` to indicate a valid status.
This is consistent with the way we use `last_refresh_error`

simon3z commented:
@yaacov @blomquisg as far as I understand *authentication* and *connection status* could be two separated statuses (columns) on the endpoints. Do we really need this distinction?

@blomquisg I assume that there's a lot of code dealing with the *authentication* value (retries, etc.) so maybe to extend that to represent also the *connection status* would be tricky.

On the other hand, again, having two different fields to represent them seems an overkill.

moolitayer commented on db/migrate/20160808150745_add_metrics_status_to_ext_management_system.rb:
```diff
@@ -0,0 +1,7 @@
+class AddMetricsStatusToExtManagementSystem < ActiveRecord::Migration[5.0]
+  def change
+    add_column :ext_management_systems, :last_metrics_error, :text
+    add_column :ext_management_systems, :last_metrics_update_date, :timestamp
```
`metrics_state_update_date` ?

moolitayer commented on db/migrate/20160808150745_add_metrics_status_to_ext_management_system.rb:
```diff
@@ -0,0 +1,7 @@
+class AddMetricsStatusToExtManagementSystem < ActiveRecord::Migration[5.0]
+  def change
+    add_column :ext_management_systems, :last_metrics_error, :text
```
This name is bad since it can contain 'Invalid', 'Unavailable' or 'Valid' which is not an error. I suggest: `metrics_state`  - the word state implies 'last'.

blomquisg replied:
:+1:, though the `last_` part would keep it consistent with the `last_refresh` columns.

Btw, we used `last_refresh_error` to store the error value, and an empty `last_refresh_error` indicated that `last_refresh_date` was the date of the last *successful* refresh.

I'm fine with tracking metrics with three columns showing status, last updated, last success.

We can settle on the correct list of fields for a generic refactoring later, I think.

blomquisg commented:
I'm still going through the `textual_` methods making sure I didn't miss anything...almost done.

blomquisg commented:
Last check, I'm having @jameswnl check if this feature has any impact on other providers as is.

As long as other providers largely ignore this feature for now, then I'm good with merging it.  Then, we can come back and extend other providers to implement this functionality as well.

blomquisg commented:
While jameswnl is looking over the impact to other providers, let's have one more go around with the field names to make sure we're all good here.

* `last_metrics_error` ... since you're storing success and failure status values here, I suggest this be `last_metrics_status`
* `last_metrics_update_date` ... I'm good with this name ... naming is hard, and there's probably a better way to call this, but I can settle with this one
* `last_metrics_success_date` ... same, good with this name

Anyone else have any input here?  @moolitayer?

jameswnl commented:
Ok, it doesn't seem to interfere with at least VMware providers. 

I am able to perform a C&U capture for the 2 providers in my setup and the ```ext_management_systems``` table for them still have the 3 newly added columns empty. And summary page only shows ```Last Refresh```

yaacov commented:
> let's have one more go around with the field names to make sure we're all good here.

> last_metrics_error ... since you're storing success and failure status values here ...

@blomquisg Changed `last_metrics_error` to only store the last error message (as symbol) or empty to indicate success.

moolitayer commented:
LGTM :+1: 
nilable ```last_metrics_error``` now seems consistent with last_refresh_error
cc @blomquisg 

@yaacov @blomquisg do we have an issue to track infra changes discussed here?

blomquisg commented:
> do we have an issue to track infra changes discussed here?

Yeah ... I created something on our pivotal board.

yaacov commented:
@simon3z @blomquisg 
This PR has a bug for 5.6.2
https://bugzilla.redhat.com/show_bug.cgi?id=1356130
Is this going to be backported ?

moolitayer commented:
Reached agreement on 1356130 NOT to backport this.
@miq-bot remove_label darga/yes
@miq-bot add_label darga/no
